pingcap/tiflow
# pulsar (ticdc): fix tls encryption and oauth2.0

### What problem does this PR solve?


Issue Number: close #10602 

### What is changed and how it works?

1. Modify the `dmlSinkFactory` to support `PulsarSSLScheme`.
2. Enable the use of the OAuth2.0 authentication function of the pulsar sink without requiring TLS configuration, as they are unrelated.
3. Enable the TLS encryption function of the pulsar sink functions correctly.

### Check List 

#### Tests 

 - Manual test (add detailed scripts or steps below)
    1. Generate certificate/key pairs for the CA, broker, and client for the Pulsar server and client. Follow the instructions at https://pulsar.apache.org/docs/2.11.x/security-tls-transport/.
    2. Obtain the OAuth 2.0 related credentials from https://auth0.com/.
    3. Deploy a standalone Pulsar instance with SSL encryption and OAuth 2.0 authentication enabled. Add the necessary configuration to the end of the conf/standalone.conf file in the Pulsar directory.
        
        ```bash
        # configure TLS ports
        brokerServicePortTls=6651
        webServicePortTls=8081
        
        # configure CA certificate
        tlsTrustCertsFilePath=./my-ca/certs/ca.cert.pem
        # configure server certificate
        tlsCertificateFilePath=./my-ca/broker.cert.pem
        # configure server's private key
        tlsKeyFilePath=./my-ca/broker.key-pk8.pem
        
        # Tls cert refresh duration in seconds (set 0 to check on every new connection)
        tlsCertRefreshCheckDurationSec=300
        
        # below is the oauth2 authentication configuration
        authenticationEnabled=true
        authenticationProviders=org.apache.pulsar.broker.authentication.AuthenticationProviderToken
        brokerClientAuthenticationPlugin=org.apache.pulsar.client.impl.auth.oauth2.AuthenticationOAuth2
        brokerClientAuthenticationParameters={\"type\":\"client_credentials\",\"privateKey\":\"./apache-pulsar-2.11.0/oauth2-conf/client_credentials.json\",\"issuerUrl\":\"https://dev-xxx(masked).us.auth0.com\", \"audience\": \"pulsar\"}
        tokenSecretKey=data:;base64,xxxx # The secret key is masked.
        ```
        
    4. Create a changefeed configuration file named `pulsar-oauth2.toml` with the following content:
        
        ```bash
        [sink.pulsar-config]
        # Pulsar oauth2 issuer-url more detail see：https://pulsar.apache.org/docs/2.10.x/client-libraries-go/#oauth2-authentication
        oauth2.oauth2-issuer-url="https://dev-xxxx(masked).us.auth0.com"
        # Pulsar oauth2 audience
        oauth2.oauth2-audience="pulsar"
        # Pulsar oauth2 private-key
        oauth2.oauth2-private-key="./client_credentials.json"
        # Pulsar oauth2 client-id
        oauth2.oauth2-client-id="xxxx" # masked
        
        # tls ca file path, configure for ssl enscription 
        tls-trust-certs-file-path="/Users/dongmen/dongmen/works/apache-pulsar-2.11.0/my-ca/certs/ca.cert.pem"
        ```
        
        ```bash
        ./cdc cli changefeed create --server=127.0.0.1:8300 --changefeed-id="pulsar-ssl-test"  --sink-uri="pulsar+ssl://127.0.0.1:6651/consumer-test?protocol=canal-json" --config pulsar-oauth2.toml
        
        Create changefeed successfully!
        ID: pulsar-ssl-test
        Info: ...
        ```
    5. Execute some DDLs and DMLs in the upstream TiDB cluster and everything works well.

#### Questions 

##### Will it cause performance regression or break compatibility?

##### Do you need to update user documentation, design documentation or monitoring documentation?

### Release note 

```release-note
Fix TLS encryption and Oauth2.0 authentication function for Pulsar sink.
```
